Feature/change slug after update api name

saransh-dev opened the pull request:
Closes #2805

### Changes
apis/client/profile/settings/details/autoform.js
apis/client/profile/view.js
apis/collection/slugs.js

marla-singer commented on apis/client/profile/view.js:
```diff
@@ -20,18 +21,18 @@ import ApiDocs from '/api_docs/collection';
 Template.viewApi.onCreated(function () {
   // Get reference to template instance
   const instance = this;
+  Tracker.autorun(() => {
```
Use `instance` variable instead of `Tracker`

saransh-dev replied:
We are using `instance` for `this` i.e., we are provide the scope of `Template` in the `instance` variable. But `Tracker` is the package of `meteor.

marla-singer replied:
@saransh-dev We don't use `autorun` function of Tracker library inside `Template`. We're using `autorun` function of Template

http://blazejs.org/api/templates.html#Blaze-TemplateInstance-autorun

brylie replied:
E.g.

```js
Template.templateName.onCreated(function () {
  // Get reference to template instance
  const templateInstance = this;

  // attach autorun to template instance
  templateInstance.autorun(function () {
    // reactive code
  });
});
```

Pardon any semantic or syntactic errors above, as I am writing this from memory.

saransh-dev replied:
I changed code as you suggested but it does not work properly
Like:- 
if we use below code
```javascript
const instance = this;
  instance.autorun(function(){

    instance.slug = FlowRouter.getParam('slug');

    // Subscribe to API and related organization
    instance.subscribe('apiComposite', instance.slug);

    // Subscribe to public proxy details
    instance.subscribe('proxyCount');

    // Subscribe to public proxy details for proxy form
    instance.subscribe('publicProxyDetails');
  });
});
```

it redirect to `/apis` because `api` gets undefined in `const api = Apis.findOne(this.docId)` in below code
```javascript
AutoForm.hooks({
  apiDetailsForm: {
    onSuccess () {
        const api = Apis.findOne(this.docId);
        if (api && api.slug) {
        
        // Redirect to updated API with new slug. It is use while api's name will update slug have to change that's mean routing change
        FlowRouter.go('viewApi', { slug: api.slug });
      } else {
        // Otherwise Redirect to API Catalog
        FlowRouter.go('apiCatalog');
      }
      // Get success message translation
      const message = TAPi18n.__('apiDetailsForm_text_updateInformation');

      // Show message
      sAlert.success(message);
    },
  },
});
```


And if I use `Tracker` then it gives results.

marla-singer replied:
@saransh-dev 

Try an arrow function:
```js
const instance = this;
  instance.autorun(() => { ...});
```

brylie replied:
I think the problem is not telated to using an arrow function, since the autorun function does not need to share a lexical `this` with its outer context.

Double check what `instance` contains. Specifically, note whether `instance.slug` exists (semantic bug).

brylie replied:
The reason we use `templateInstance` instead of just `instance` is so it is explicit that we are referring to a **template instance**. In other words, the variable name `instance` is ambiguous, and may contribute to confusion when reasoning about the code.

brylie replied:
It may also be that the `templateInstance.autorun` pattern is irrelevant here, since we are in an AutoForm onSuccess handler.

The arguments or data context of AutoForm success handler may differ depending on whether we are inserting or updating a document.

brylie replied:
It may also be that the `templateInstance.autorun` pattern is irrelevant here, since we are in an AutoForm onSuccess handler.

The arguments or data context of AutoForm success handler may differ depending on whether we are inserting or updating a document.

marla-singer replied:
@brylie After debugging I found that `const api = Apis.findOne(this.docId);` returns `undefined`. I can't understand the reason :thinking: So the problem isn't related to `autorun` function

saransh-dev replied:
If i will update anything instead of api's name then it work perfectly but when i update api's name, it redirects me on `/apis` because `Apis.findOne(this.docId);` returns `undefined`.

brylie replied:
As a basic troubleshooting step, what does `this` refer to in your above example?

marla-singer replied:
@brylie `this.docId` returns a correct ID value of updated document

saransh-dev replied:
@marla-singer yes, `this.docId` return correct id of updated record but `Apis.findOne(this.docId);` returns `undefined` while updating api's name.

marla-singer commented on apis/client/profile/settings/details/autoform.js:
```diff
@@ -7,11 +7,20 @@ https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence
       // Get success message translation
+      if (this.updateDoc && this.updateDoc.$set.name) {
+        const slug = this.updateDoc.$set.name.split(' ').join('-').toLowerCase();
+        // Redirect to newly added API
```
Why did you do this? As I remember `slug` is auto-generated

brylie replied:
I agree with  @marla-singer that we should use the auto-generated slugs if possible. This is due to the necessary **uniqueness of slugs**, where our auto-slug will append a number to ensure slugs are unique.

saransh-dev replied:
Yes, slug is auto-generated but while update we get collection's id inside this.docId and all record which will be updated inside this.updateDoc. And we can not able to find collection on the base of collection's Id for getting slug. Slug is auto generated and it always update while api's name will update. On line 17, i wrote a code to get slug from api's name by using some javascript function and that slug is using in line 19, pass in the params inside FlowRouter.

marla-singer replied:
Check docs of autoform
https://github.com/aldeed/meteor-autoform#callbackshooks

It has to be ID of API document as the second parameter after that you can get API instance from database and get slug value.

Remake

saransh-dev replied:
ok i got your point and i'll change it and commit the code.

marla-singer commented on apis/client/profile/settings/details/autoform.js:
```diff
@@ -7,11 +7,20 @@ https://joinup.ec.europa.eu/community/eupl/og_page/european-union-public-licence
+        const slug = this.updateDoc.$set.name.split(' ').join('-').toLowerCase();
+        // Redirect to newly added API
+        FlowRouter.go('viewApi', { slug });
+      } else {
```
And why did you do redirect in this line?

brylie replied:
The `FlowRouter.go()` line may be needed because changing the slug should take effect in the URL. If we don't redirect, the old slug would still be visible in the URL.

saransh-dev replied:
I apology for the comment (on line 18). That comment is wrong, that's why it creates some confusion.
It is used to redirect to updated API with a new slug i.e., while API's name will update slug have to change that's mean routing change

marla-singer commented:
@saransh-dev Squash your commits to one

marla-singer commented:
@saransh-dev PR is not ready to review

1. You have ESlint error

2. The realisation with `Tracker` is wrong

## Steps for reproduce
1. Open a browser console
2. Go to APIs catalog
3. Select an API
4. GO back to catalog

## Found
`
Exception from Tracker recompute function:
Error: Can't call View#subscribe from inside the destroyed callback, try calling it inside created or rendered.
  at Blaze.View._errorIfShouldntCallSubscribe (blaze.js?hash=f33d3df…:1965)
  at Blaze.View.subscribe (blaze.js?hash=f33d3df…:1979)
  at Blaze.TemplateInstance.subscribe (blaze.js?hash=f33d3df…:3682)
  at view.js:28
`
`Tracker.autorun()` runs inside the destroyed template`viewApi`

3. After updating name it goes to new URL (that's expected) but a browser console has a lot of errors

saransh-dev commented:
Hi @marla-singer , as per my understanding `Apis.findOne(this.docId);` returns `undefined` because in view.js file, there is a subscription which subscribes the single API data by slug and when we update the API's name then slug change but because of current route, old slug goes in subscription and no record publish and we can not able to find API record by id.

I will give you short example:-

Let we have URL `apis/test-api`
where `test-api` is slug and API's name is `Test API`.

here is the subscription:-

```javascript
 		instance.slug = FlowRouter.getParam('slug');

    // Subscribe to API and related organization
    instance.subscribe('apiComposite', instance.slug)
```

In above code, we are passing slug in the subscription.

Now I am going to update API name to `Test`. Now slug name will be `test`. But because of current route subscription takes old slug which was `test-api` and no data record publish and inside autoform hook `Apis.findOne(this.docId);` returns `undefined` because before passing new slug in subscription we find the record for getting slug name and will get `undefined` because no record will publish.

For finding subscription problem I have added `autopublish` package in app and comment above subscription code and after update API name and `Apis.findOne(this.docId);` returns data with respective id.

For this we can use slug name directly by converting API name through some javascript function like below:- 
`this.updateDoc.$set.name.split(' ').join('-').toLowerCase()`

And we have one more way like:-
 
We can make a server side method to find API by Id and return record and call that method in `onSuccess()` inside autoform hook and call routing inside Meteor.call's callback.


Please advise.

Thanks

marla-singer commented:
> For this we can use slug name directly by converting API name through some javascript function like below:-
`this.updateDoc.$set.name.split(' ').join('-').toLowerCase()`

It is a narrow solution and will work correctly only with English API name. How will it work with not the Latin alphabet?

> And we have one more way like:-
We can make a server side method to find API by Id and return record and call that method in onSuccess() inside autoform hook and call routing inside Meteor.call's callback.

Or you take a new slug using `this.updateDoc.$set.slug` and it will be the most right choice. (You can test and make sure `this.updateDoc.$set.slug` has an updated value of slug)

And said above the problem comes because of subscription. Even if you roll back to your previous solution with using `this.updateDoc.$set.name.split(' ').join('-').toLowerCase()` then you will see the errors in browser console

brylie commented:
@saransh-dev we should rely on the *todda00:friendly-slugs* package, which is already active in our project. Please **do not** write your own slug generating code, as there are too many edge cases (such as unicode and uniqueness).

brylie commented:
An alternate solution might be:

- create a Meteor Method called `getApiSlugById(apiId)` --> `apiSlug`
- in the AutoForm `onSucces` hook, call the `getApiSlugById` method with the current `apiId`
- in the `getApiSlugById` callback function, route to the API profile using the returned `apiSlug`

saransh-dev commented:
@brylie and @marla-singer, Thanks and I am going to implement alternate solution `Meteor Method called getApiSlugById(apiId)`

saransh-dev commented:
Hi @brylie and @marla-singer  , I have faced one more issue. I have implemented `Meteor.call` and its work fine but there is one more thing we need to take care about that.

When we update the API's name and slug also updated and redirect on the new slug. In the middle of this process update and redirect, some exception occurs. 

```javascript
Exception in template helper: TypeError: Cannot read property '_id' of undefined
    at Object.api (http://localhost:3000/app/app.js?hash=35532a307fe139b5c5f366ae17d51fb9de01a130:9071:25)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3051:16
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1715:16
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3103:66
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3744:12)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3102:27
    at Object.Spacebars.call (http://localhost:3000/packages/spacebars.js?hash=ebf9381e7fc625d41acb0df14995b7614360858a:172:18)
    at http://localhost:3000/app/app.js?hash=35532a307fe139b5c5f366ae17d51fb9de01a130:8865:24
    at wrappedArgFunc (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:2960:14)
    at Blaze.View.<anonymous> (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:2673:26)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1934:20
    at Function.Template._withTemplateInstanceFunc (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:3744:12)
    at http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1932:29
    at Object.Blaze._withCurrentView (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:2271:12)
    at viewAutorun (http://localhost:3000/packages/blaze.js?hash=f33d3dfed63a491d24e3aa07ad66c24b5fe8c761:1931:18)
    at Tracker.Computation._compute (http://localhost:3000/packages/tracker.js?hash=997515fa2d5b0530ba07741da556c4b36963ef3b:339:36)
    at Tracker.Computation._recompute (http://localhost:3000/packages/tracker.js?hash=997515fa2d5b0530ba07741da556c4b36963ef3b:358:14)
    at Tracker._runFlush (http://localhost:3000/packages/tracker.js?hash=997515fa2d5b0530ba07741da556c4b36963ef3b:532:14)
    at e.exports (http://localhost:3000/swagger/bundle.js:1:8056)
    at w.(anonymous function) (http://localhost:3000/swagger/bundle.js:105:7780)
    at Number.S (http://localhost:3000/swagger/bundle.js:105:7638)
    at MessagePort.E (http://localhost:3000/swagger/bundle.js:105:7660)
 ```

 It is so because when we update the API's name, slug change after this and before redirecting new slug, we have old slug in `instance.slug` and `FlowRouter.getParams('slug')`

 And we have following function inside `helpers` inside `view.js`

 ```javascript
api () {
    // Get reference to template instance
    const instance = Template.instance();

    // Get API slug
    const slug = instance.slug;

    // Get single API Backend
    const api = Apis.findOne({ slug });
    
    // Save the API ID
    instance.apiId = api._id;

    return api;
},
```

In `const api = Apis.findOne({ slug });` returns `undefined` because of new slug and we also set `api._id` in `instance.apiID`. But in that line it thorws exception `Exception in template helper: TypeError: Cannot read property '_id' of undefined` it means `api` is `undefined`.

For this, we need to make some checks in `helpers`.

Please advise.

Thanks.

brylie commented:
Would this resolve the console error?

```js
Template.apiProfile.onCreated(function () {
  // Get reference to template instance
  const templateInstance = this;

  // Reactive variable to hold API document
  templateInstance.api = new ReactiveVar();
  
  templateInstance.autorun(function () {
   // Get API slug (whichever name the URL parameter has)
   const slug = FlowRouter.getParam("apiSlug");

   // Get single API Backend
   const api = Apis.findOne({ slug });

    if (api) {
      templateInstance.api.set(api);
    }
  });
});
api () {
   
Template.apiProfile.helpers({
  api () {
    // NOTE: keep helpers simple and focused

    // Get reference to template instance
    const templateInstance = Template.instance();

    // Get API from reactive variable
   const api = templateInstance.api.get();

    return api;
  }
});
},
```

marla-singer commented:
> in the AutoForm onSucces hook, call the getApiSlugById method with the current apiId

@brylie I totally disagree about this double routine! As I wrote above, the variable `this.updateDoc.$set` in the AutoForm onSuccess hook **already has** a relevant value of slug -->
```json
{"name":"Tested API","url":"http://url.com","slug":"tested-api","updated_at":"2017-08-25T08:13:00.822Z"}
```

brylie commented:
> I totally disagree about this double routine!

Kindly post some proposed code.

marla-singer commented:
Instead of (and using Meteor method call)

```js
AutoForm.hooks({
  apiDetailsForm: {
    onSuccess () {
        const api = Apis.findOne(this.docId);
        if (api && api.slug) {
        
        // Redirect to updated API with new slug. It is use while api's name will update slug have to change that's mean routing change
        FlowRouter.go('viewApi', { slug: api.slug });
      } else {
        // Otherwise Redirect to API Catalog
        FlowRouter.go('apiCatalog');
      }
      // Get success message translation
      const message = TAPi18n.__('apiDetailsForm_text_updateInformation');

      // Show message
      sAlert.success(message);
    },
  },
});
```

Use
```js
AutoForm.hooks({
  apiDetailsForm: {
    onSuccess () {
        const slug = this.updateDoc.$set.slug;
        // make sure slug was edited
        if (slug) {
        // Redirect to updated API with new slug
        FlowRouter.go('viewApi', { slug });
      }

      // Get success message translation
      const message = TAPi18n.__('apiDetailsForm_text_updateInformation');

      // Show message
      sAlert.success(message);
    },
  },
});
```

brylie commented:
Also, please resolve conflicts.

saransh-dev commented:
Hi @brylie , I have faced on more exception inside `Template.viewApiPageHeader.helpers` in `header.js` file in below function:-


```javascript
userShouldSeeBacklogTab () {
  // Get API id
  const apiId = this.api._id;
  console.log(':: apiId viewApiPageHeader -> ',apiId)
  const api = Apis.findOne(apiId);
  console.log(':: api viewApiPageHeader -> ',api,' -> this.api ',this.api)
  // Check if API Backlog exist or user allowed to see
  if (api.backlogIsNotEmpty() || api.currentUserCanManage()) {
    return true;
  }
  return false;
},
```

After update api's name, `api` inside `console.log(':: api viewApiPageHeader -> ',api,' -> this.api ',this.api)` retruns `undefined` first time and second time it returns data. But in `this.api` have data with old api's name at first time but on second time `this.api` have updated data.

Because of " `api` returns `undefined` ", some exceptions occur in each function of this helpers.

Can we use `this.api` instead of `Apis.findOne(apiId)` ?

Please advise.

bajiat commented:
@marla-singer or  @matleppa Could you help @saransh-dev ?

brylie commented:
> Because of " api returns undefined ", some exceptions occur in each function of this helpers.

We have to make sure an object exists before using a property of the object:

```js
if (object) {
  // use object property
  const property = object.property;
}
```

This is a recurring source of bugs, with a relatively simple fix.

brylie commented:
Please resolve conflicts.

brylie commented:
You have too many files as part of this PR. Create a new branch from `develop` and cherry-pick the commits you made from this PR. Then, close this PR and delete this branch.

saransh-dev commented:
@brylie Do i delete this branch and create new branch with same name?
